eth/rpc/backend.go

```diff
@@ -200,25 +190,35 @@ func (b *backend) HeaderByHash(_ context.Context, hash common.Hash) (*types.Head
 func (b *backend) HeaderByNumberOrHash(_ context.Context,
 	blockNrOrHash BlockNumberOrHash,
 ) (*types.Header, error) {
-	block, err := b.polarisBlockByNumberOrHash(blockNrOrHash)
-	// If the block is non-existent, return nil.
-	// This is to maintain parity with the behavior of the geth backend.
-	if errors.Is(err, core.ErrBlockNotFound) {
-		return nil, nil //nolint:nilnil // we love go-ethereum.
+	if blockNr, ok := blockNrOrHash.Number(); ok {
+		block := b.polarisBlockByNumber(blockNr)
+		if block == nil {
+			return nil, errors.New("header for hash not found")
```

Comment:
should be "header for number not found"

Also, if you cant find by number, you should try finding by hash right. hence the byNumberORHash